### PR TITLE
Add comparator to track difference before and after

### LINKED ISSUE #96: HMI shall not send a OnInteriorVehicleData notification if a parameter is included in the request but the value is the same as the current value
HMI shall not send a OnInteriorVehicleData notification if a parameter is included in the request but the value is the same as the current value (i.e. no change).
(This issue is applicable to all types of modules)

Steps: 
1. mobile sends a valid SetInteriorVehicleData to HMI with some params that will be changed 
2. HMI apply that changes and send OnInteriorVehicleData notification with changed parameters
3. mobile sends a valid SetInteriorVehicleData to HMI **_with the same params_**

Expected:
HMI doesn't send OnInteriorVehicleData notification because the params values are the same as the current values

Observed:
HMI send OnInteriorVehicleData notification

### PR DESCRIPTION
**Pull request by AKalinich-Luxoft**
And send OnIVD notification only for case when some fields has
actually been changed.

Fixes #96 

This PR is **ready** for review.

### Testing Plan
Steps: 
1. mobile sends a valid SetInteriorVehicleData to HMI with some params that will be changed 
2. HMI apply that changes and send OnInteriorVehicleData notification with changed parameters
3. mobile sends a valid SetInteriorVehicleData to HMI **_with the same params_**

### Summary
Add comparator to track difference before and after SetIVD request for each module data and send OnIVD if some values were actually changed

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: app/controller/sdl/RCModulesController.js [hunk 1]
```diff
@@ -923,12 +946,26 @@ SDL.RCModulesController = Em.Object.create({
...
-                                                                lightControlData: data.params.moduleData.lightControlData});
+                    data_before_tmp.lightState.forEach( (element_dest) => {
+                      lightControlData.lightState.forEach( (element_target) => {
+                        if (element_target.id == element_dest.id) {
+                          data_before.lightState.push(element_dest);
+                        }
+                      });
+                    });
+
+                    const changedFields = getDataDifference(data_before, lightControlData);
```
Comment (ShobhitAd) [inline app/controller/sdl/RCModulesController.js:965]: This doesn't work correctly if only some properties are changed. For example sending `SetInteriorVehicleData` twice with params

```
{
	"moduleData": {
		"moduleType": "LIGHT",
		"lightControlData": {
			"lightState": [{
				"id": "FRONT_LEFT_HIGH_BEAM",
				"status": "ON",
				"density": 1
			}]
		}
	}
}
```
results in two `OnInteriorVehicleData` notifications being sent from the HMI. Probably due to extra params (like `color`) not being ignored by the [`if (undefined !== another[param] &&`](https://github.com/LuxoftSDL/sdl_hmi/blob/fix/fix_on_ivd_notification_after_setivd/app/controller/sdl/RCModulesController.js#L791) check.

**Review by ShobhitAd — CHANGES_REQUESTED** [review]

File: app/controller/sdl/RCModulesController.js (hunk 1 above)
Comment (AKalinich-Luxoft) [reply app/controller/sdl/RCModulesController.js:965]: @ShobhitAd are you sure you are sending SetIVD twice with the same params? When I am trying to send the same params like you described above, I am getting only one OnIVD notification:
```
SDL -> HMI [14:10:59:485]: {"id":50,"jsonrpc":"2.0","method":"RC.SetInteriorVehicleData","params":{"appID":1111048898,"moduleData":{"lightControlData":{"lightState":[{"color":{"blue":0,"green":0,"red":0},"density":0,"id":"FRONT_LEFT_HIGH_BEAM","status":"ON"}]},"moduleId":"d38e4a05-b17c-28e3-9d38-e4a05b17c28e","moduleType":"LIGHT"}}}

RCRPC.js:225 FFW.RC.SetInteriorVehicleData Request
RCRPC.js:529 FFW.RC.OnInteriorVehicleData Notification
RPCClient.js:323 HMI -> SDL [14:12:17:114]: {"jsonrpc":"2.0","method":"RC.OnInteriorVehicleData","params":{"moduleData":{"moduleType":"LIGHT","moduleId":"d38e4a05-b17c-28e3-9d38-e4a05b17c28e","lightControlData":{"lightState":[{"color":{"blue":0,"green":0,"red":0},"density":0,"id":"FRONT_LEFT_HIGH_BEAM","status":"ON"}]}}}}
RPCClient.js:323 HMI -> SDL [14:12:17:114]: {"jsonrpc":"2.0","id":50,"result":{"code":0,"method":"RC.SetInteriorVehicleData","moduleData":{"moduleType":"LIGHT","moduleId":"d38e4a05-b17c-28e3-9d38-e4a05b17c28e","lightControlData":{"lightState":[{"color":{"blue":0,"green":0,"red":0},"density":0,"id":"FRONT_LEFT_HIGH_BEAM","status":"ON"}]}}}}
RPCClient.js:137 SDL -> HMI [14:12:23:144]: {"id":54,"jsonrpc":"2.0","method":"RC.SetInteriorVehicleData","params":{"appID":1111048898,"moduleData":{"lightControlData":{"lightState":[{"color":{"blue":0,"green":0,"red":0},"density":0,"id":"FRONT_LEFT_HIGH_BEAM","status":"ON"}]},"moduleId":"d38e4a05-b17c-28e3-9d38-e4a05b17c28e","moduleType":"LIGHT"}}}

RCRPC.js:225 FFW.RC.SetInteriorVehicleData Request
RPCClient.js:323 HMI -> SDL [14:16:6:789]: {"jsonrpc":"2.0","id":54,"result":{"code":0,"method":"RC.SetInteriorVehicleData","moduleData":{"moduleType":"LIGHT","moduleId":"d38e4a05-b17c-28e3-9d38-e4a05b17c28e","lightControlData":{"lightState":[{"color":{"blue":0,"green":0,"red":0},"density":0,"id":"FRONT_LEFT_HIGH_BEAM","status":"ON"}]}}}}
```

Comment (ShobhitAd) [reply app/controller/sdl/RCModulesController.js:965]: @AKalinich-Luxoft I re-ran the test but I'm still seeing the same issue. I'm attaching my HMI console logs for the test
[OnIVD_LIGHT.log](https://github.com/smartdevicelink/sdl_hmi/files/7079405/OnIVD_LIGHT.log)

I think the issue occurs when one or more optional params are omitted from the request. It looks like if `compareObjects` is called with extra params in the second argument https://github.com/smartdevicelink/sdl_hmi/blob/master/app/controller/sdl/Abstract/Controller.js#L408.

Could you retry the test you mentioned and send the `SetInteriorVehicleData` requests without the `color` parameter(only send the `density`, `id` and `status` params)

Comment (AKalinich-Luxoft) [reply app/controller/sdl/RCModulesController.js:965]: @ShobhitAd thank you for the additional input, I see now. I also noticed that my SPT sends optional params even if I unchecked them in UI.
Looks like we should consider the field as changed only when the comparator returns positive 1. I updated the condition and now it works as expected. See https://github.com/smartdevicelink/sdl_hmi/pull/593/commits/553aebaa059c5a98c8f635197eccd9d38255a8f6

**Review by ShobhitAd — APPROVED** [review]

**Pull request merged**